actioncable/lib/action_cable/subscription_adapter/postgresql.rb

```diff
@@ -29,15 +29,20 @@ def shutdown
       end
 
       def with_connection(&block) # :nodoc:
-        ActiveRecord::Base.connection_pool.with_connection do |ar_conn|
-          pg_conn = ar_conn.raw_connection
-
-          unless pg_conn.is_a?(PG::Connection)
-            raise "The Active Record database must be PostgreSQL in order to use the PostgreSQL Action Cable storage adapter"
-          end
+        ar_conn = ActiveRecord::Base.connection_pool.checkout.tap do |conn|
+          # Action Cable is taking ownership over this database connection, and
+          # will perform the necessary cleanup tasks
+          ActiveRecord::Base.connection_pool.remove(conn)
+        end
 
-          yield pg_conn
+        pg_conn = ar_conn.raw_connection
+        unless pg_conn.is_a?(PG::Connection)
+          raise "The Active Record database must be PostgreSQL in order to use the PostgreSQL Action Cable storage adapter"
         end
```

Comment:
Even ignoring the difficulty of checking earlier, that would seem pretty defensive to me.

Here, its purpose is more to translate the error you'd get from the block into something immediately obvious -- it's just easier to check the condition first instead of using a rescue.